src/lib/decorators.ts

```diff
@@ -225,6 +225,54 @@ export function query(selector: string) {
   };
 }
 
+/**
+ * A property decorator that converts a class property into a getter that
+ * returns a promise that resolves to the result of a querySelector on the
+ * element's renderRoot done after the element's `updateComplete` promise
+ * resolves and after the queried element's `updateComplete` promise resolves
+ * (if it is a LitElement).
+ *
+ * @param selector A DOMString containing one or more selectors to match.
+ *
+ * See: https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector
+ *
+ * @example
+ *
+ *     class MyElement {
+ *       @asyncQuery('#first')
+ *       first;
+ *
+ *       render() {
+ *         return html`
+ *           <div id="first"></div>
+ *           <div id="second"></div>
+ *         `;
+ *       }
+ *     }
+ */
+export function asyncQuery(selector: string) {
+  return (protoOrDescriptor: Object|ClassElement,
+          // tslint:disable-next-line:no-any decorator
+          name?: PropertyKey): any => {
+    const descriptor = {
+      async get(this: LitElement) {
+        await this.updateComplete;
+        const el = this.renderRoot.querySelector(selector);
+        // if this is a LitElement, then await updateComplete
+        if (el) {
+          await (el as LitElement).updateComplete;
```

Comment:
Is this good/required?

It sort of confuses the question "why should a user use `asyncQuery`?"
1. Because you need a reference to a child element but it won't be there until after _this_ element flushes? 
1. Because you've changed state on the _child_ element and for some reason don't want to interact with it until after it's flushed?

The latter should be rare; flushing a child should normally be unobservable except for measurement (and we know that's fraught due to nested/chained updates). It's sort of lumping two use cases together, and given the second one is problematic, just want to discuss whether the child await is defensible vs. just pushing the entire measurement case to user responsibility.

Reply:
This also causes another microtask wait for every use, regardless of whether the element is a LitElement. And it doesn't work for other asynchronously rendering components like Stencil.

Reply:
Yup, removing this part.

Reply:
I think this is potentially very useful if made a bit more flexible:

```ts
export function asyncQuery(selector: string, waitFor?: (el) => Promise<unknown>) {
  if (waitFor) { await waitFor(el); }
...
```

Then one could do:

```ts
@asyncQuery('#foo', (e) => e.updateComplete) foo;
```

Or the equivalent for Stencil...

Reply:
Added a comment about supporting this in the future.